# Topic 9 — Pandas Data Manipulation — 📖 Lesson

## From "I know DataFrames" to confident, professional data wrangling

> **Analogy:** If a DataFrame is a spreadsheet, data manipulation is everything you do with the spreadsheet's tools: **select**, **filter**, **sort**, **add formula columns**, **group & summarise**, **pivot**, **clean text and dates** — but with code that is fast, repeatable and scales to millions of rows.

### Roadmap

| Part | Sections |
|---|---|
| 🟢 **Everyday operations** | 1 Selecting · 2 Filtering · 3 Adding/modifying columns · 4 Sorting · 5 Renaming · 6 Dropping · 7 Index |
| 🟡 **Transforming & summarising** | 8 apply/map · 9 GroupBy · 10 transform · 11 filter · 12 Pivot tables · 13 Crosstab |
| 🟡 **Cleaning** | 14 Strings · 15 Dates · 16 Duplicates · 17 Missing values · 18 replace/map · 19 Binning · 20 explode |
| 🔴 **Advanced** | 21 Reshaping, window & cumulative functions · 22 Performance |
| 📝 **Practice** | 20 problems → separate Solutions |
| 🚀 **Project** | E-Commerce Data Manipulation (20 steps) |
| 📋 **Reference** | Cheat sheet |

Every important concept covers: **What · Why · When · When not · Syntax · Example · Expected output · Explanation · Common mistakes**.

> ℹ️ This notebook targets **pandas 2.x/3.x**. Two modern changes are highlighted where relevant: **Copy-on-Write** (always on in pandas 3) and **`DataFrame.map`** (replaces the removed `applymap`).

> 📚 **Course navigation** — you are in **Topic 09: Pandas Data Manipulation**
> ⬅️ Previous: [Topic 08 — Data Visualization](../Topic_08_Data_Visualization/Topic_08_Lesson.ipynb)  ·  ➡️ Next: [Topic 10 — Combining DataFrames](../Topic_10_Combining_DataFrames/Topic_10_Lesson.ipynb)  ·  🗂️ [All topics](../Topic_00_All_Topics_Index.ipynb)

<details><summary>Full course sequence (click to expand)</summary>

| # | Topic | Lesson | Practice & Solutions |
|---|---|---|---|
| 01 | Python Basics | [📖 Lesson](../Topic_01_Python_Basics/Topic_01_Lesson.ipynb) | [✍️ Practice](../Topic_01_Python_Basics/Topic_01_Practice_Solutions.ipynb) |
| 02 | Pandas DataFrame Basics | [📖 Lesson](../Topic_02_Pandas_DataFrame_Basics/Topic_02_Lesson.ipynb) | [✍️ Practice](../Topic_02_Pandas_DataFrame_Basics/Topic_02_Practice_Solutions.ipynb) |
| 03 | Handling Missing Values | [📖 Lesson](../Topic_03_Handling_Missing_Values/Topic_03_Lesson.ipynb) | [✍️ Practice](../Topic_03_Handling_Missing_Values/Topic_03_Practice_Solutions.ipynb) |
| 04 | Normalization & Standardization | [📖 Lesson](../Topic_04_Normalization_Standardization/Topic_04_Lesson.ipynb) | [✍️ Practice](../Topic_04_Normalization_Standardization/Topic_04_Practice_Solutions.ipynb) |
| 05 | Detecting & Treating Outliers | [📖 Lesson](../Topic_05_Outliers/Topic_05_Lesson.ipynb) | [✍️ Practice](../Topic_05_Outliers/Topic_05_Practice_Solutions.ipynb) |
| 🎯 | *🎯 Quiz 1 — Topics 01–05* | [🎯 Quiz 1](../Quiz_1_Topics_01-05/Quiz_1.ipynb) | |
| 06 | Encoding Techniques | [📖 Lesson](../Topic_06_Encoding_Techniques/Topic_06_Lesson.ipynb) | [✍️ Practice](../Topic_06_Encoding_Techniques/Topic_06_Practice_Solutions.ipynb) |
| 07 | Exploratory Data Analysis (EDA) | [📖 Lesson](../Topic_07_Exploratory_Data_Analysis/Topic_07_Lesson.ipynb) | [✍️ Practice](../Topic_07_Exploratory_Data_Analysis/Topic_07_Practice_Solutions.ipynb) |
| 08 | Data Visualization | [📖 Lesson](../Topic_08_Data_Visualization/Topic_08_Lesson.ipynb) | [✍️ Practice](../Topic_08_Data_Visualization/Topic_08_Practice_Solutions.ipynb) |
| **▶ 09 ◀** | **Pandas Data Manipulation** | [📖 Lesson](../Topic_09_Pandas_Data_Manipulation/Topic_09_Lesson.ipynb) | [✍️ Practice](../Topic_09_Pandas_Data_Manipulation/Topic_09_Practice_Solutions.ipynb) |
| 10 | Combining DataFrames | [📖 Lesson](../Topic_10_Combining_DataFrames/Topic_10_Lesson.ipynb) | [✍️ Practice](../Topic_10_Combining_DataFrames/Topic_10_Practice_Solutions.ipynb) |
| 🎯 | *🎯 Quiz 2 — Topics 06–10* | [🎯 Quiz 2](../Quiz_2_Topics_06-10/Quiz_2.ipynb) | |
| 11 | Feature Engineering | [📖 Lesson](../Topic_11_Feature_Engineering/Topic_11_Lesson.ipynb) | [✍️ Practice](../Topic_11_Feature_Engineering/Topic_11_Practice_Solutions.ipynb) |
| 11b | KBinsDiscretizer (Discretization) | [📖 Lesson](../Topic_11b_KBinsDiscretizer/Topic_11b_Lesson.ipynb) | [✍️ Practice](../Topic_11b_KBinsDiscretizer/Topic_11b_Practice_Solutions.ipynb) |
| 12 | Correlation & Multicollinearity | [📖 Lesson](../Topic_12_Correlation_Multicollinearity/Topic_12_Lesson.ipynb) | [✍️ Practice](../Topic_12_Correlation_Multicollinearity/Topic_12_Practice_Solutions.ipynb) |
| 13 | Data Transformation & Skewness | [📖 Lesson](../Topic_13_Data_Transformation_Skewness/Topic_13_Lesson.ipynb) | [✍️ Practice](../Topic_13_Data_Transformation_Skewness/Topic_13_Practice_Solutions.ipynb) |
| 14 | Statistics for Data Science | [📖 Lesson](../Topic_14_Statistics_for_Data_Science/Topic_14_Lesson.ipynb) | [✍️ Practice](../Topic_14_Statistics_for_Data_Science/Topic_14_Practice_Solutions.ipynb) |
| 🎯 | *🎯 Quiz 3 — Topics 11–14* | [🎯 Quiz 3](../Quiz_3_Topics_11-14/Quiz_3.ipynb) | |
| 15 | Probability for Data Science & ML | [📖 Lesson](../Topic_15_Probability_for_Data_Science_ML/Topic_15_Lesson.ipynb) | [✍️ Practice](../Topic_15_Probability_for_Data_Science_ML/Topic_15_Practice_Solutions.ipynb) |
| 16 | Feature Selection | [📖 Lesson](../Topic_16_Feature_Selection/Topic_16_Lesson.ipynb) | [✍️ Practice](../Topic_16_Feature_Selection/Topic_16_Practice_Solutions.ipynb) |
| 17 | Machine Learning | [📖 Lesson](../Topic_17_Machine_Learning/Topic_17_Lesson.ipynb) | [✍️ Practice](../Topic_17_Machine_Learning/Topic_17_Practice_Solutions.ipynb) |
| 🎯 | *🏁 Final Quiz — Topics 15–17* | [🏁 Final Quiz](../Quiz_4_Topics_15-17/Quiz_4.ipynb) | |

</details>

## 📋 Lesson Index — What This Lesson Covers

This lesson has **26 sections**. Click any one to jump to it.

- [1. Selecting Data](#sec-1)
- [2. Filtering Rows](#sec-2)
- [3. Adding and Modifying Columns](#sec-3)
- [4. Sorting](#sec-4)
- [5. Renaming](#sec-5)
- [6. Dropping Data](#sec-6)
- [7. Index Manipulation](#sec-7)
- [8. Applying Functions — `apply`, `map`, `DataFrame.map` (formerly `applymap`)](#sec-8)
- [9. GroupBy (deep dive)](#sec-9)
- [10. `groupby().transform()`](#sec-10)
- [11. `groupby().filter()`](#sec-11)
- [12. Pivot Tables — `pd.pivot_table()`](#sec-12)
- [13. Crosstab — `pd.crosstab()`](#sec-13)
- [14. String Manipulation — the `.str` accessor](#sec-14)
- [15. Date/Time Manipulation](#sec-15)
- [16. Duplicate Handling](#sec-16)
- [17. Missing Value Manipulation](#sec-17)
- [18. Replace and Map](#sec-18)
- [19. Binning — `pd.cut()` and `pd.qcut()`](#sec-19)
- [20. `explode()`](#sec-20)
- [21. Advanced Pandas Manipulation](#sec-21)
- [21.1 Reshaping: MultiIndex, `stack`, `unstack`, `melt`, `pivot`, `pivot_table`, `wide_to_long`](#sec-22)
- [21.2 Conditional values: `where`, `mask`, `clip`](#sec-23)
- [21.3 Ranking and ordered calculations: `rank`, `shift`, `diff`, `pct_change`, cumulative functions](#sec-24)
- [21.4 Window functions: `rolling` and `expanding`](#sec-25)
- [22. Performance](#sec-26)

> ✍️ Practice problems with solutions: [Topic_09_Practice_Solutions.ipynb](Topic_09_Practice_Solutions.ipynb)  ·  🗂️ [All topics](../Topic_00_All_Topics_Index.ipynb)

In [1]:
import numpy as np
import pandas as pd

pd.set_option('display.width', 130)
pd.set_option('display.max_columns', 30)
print('pandas', pd.__version__, '| numpy', np.__version__)

# Main example dataset used throughout the lessons
emp = pd.DataFrame({
    'Emp_ID':     [101, 102, 103, 104, 105, 106, 107, 108, 109, 110],
    'Name':       ['Ali', 'Sara', 'Ahmed', 'Hina', 'Omar', 'Zara', 'Bilal', 'Ayesha', 'Usman', 'Fatima'],
    'Age':        [25, 32, 41, 29, 35, 27, 45, 38, 23, 31],
    'City':       ['Lahore', 'Karachi', 'Lahore', 'Islamabad', 'Karachi', 'Lahore', 'Islamabad', 'Karachi', 'Lahore', 'Islamabad'],
    'Department': ['IT', 'HR', 'IT', 'Finance', 'IT', 'HR', 'Finance', 'IT', 'Sales', 'Sales'],
    'Salary':     [85000, 60000, 120000, 90000, 98000, 55000, 130000, 110000, 45000, 70000],
    'Experience': [2, 6, 15, 5, 9, 3, 20, 12, 1, 7],
})
emp

pandas 3.0.3 | numpy 2.4.6


,Emp_ID,Name,Age,City,Department,Salary,Experience
0,101,Ali,25,Lahore,IT,85000,2
1,102,Sara,32,Karachi,HR,60000,6
2,103,Ahmed,41,Lahore,IT,120000,15
3,104,Hina,29,Islamabad,Finance,90000,5
4,105,Omar,35,Karachi,IT,98000,9
5,106,Zara,27,Lahore,HR,55000,3
6,107,Bilal,45,Islamabad,Finance,130000,20
7,108,Ayesha,38,Karachi,IT,110000,12
8,109,Usman,23,Lahore,Sales,45000,1
9,110,Fatima,31,Islamabad,Sales,70000,7


---
<a id="sec-1"></a>
# 1. Selecting Data

### What & why
Picking the **columns** and/or **rows** you need — the most common operation in any analysis.

### Syntax
| Syntax | Returns | Meaning |
|---|---|---|
| `df['col']` | **Series** | One column |
| `df[['col1', 'col2']]` | **DataFrame** | Several columns (note the **double brackets** = a list) |
| `df.loc[rows, cols]` | Series/DataFrame/value | Select by **labels** (index labels & column names) |
| `df.iloc[rows, cols]` | Series/DataFrame/value | Select by **integer positions** (0, 1, 2 …) |

### `loc` vs `iloc` — the key difference
| | `loc` | `iloc` |
|---|---|---|
| Uses | **Labels** (names) | **Positions** (numbers) |
| Slice end | **Included**: `loc[0:3]` → labels 0, 1, 2, **3** | **Excluded**: `iloc[0:3]` → positions 0, 1, 2 |
| Accepts boolean masks | ✅ `loc[df['Age'] > 30]` | Only arrays, not Series |
| Column selection | By name: `loc[:, 'Salary']` | By number: `iloc[:, 5]` |

**Memory trick:** **l**oc = **l**abel, **i**loc = **i**nteger.

With the default index (0, 1, 2, …) labels and positions look the same — the difference becomes obvious once the index is something else (e.g. after sorting or `set_index`).

In [2]:
print(type(emp['Name']), '← single brackets → Series')
print(type(emp[['Name', 'Salary']]), '← double brackets → DataFrame\n')

print('loc[0:3, ["Name","Age"]] → labels 0..3 INCLUDED')
print(emp.loc[0:3, ['Name', 'Age']], '\n')
print('iloc[0:3, [1, 2]] → positions 0..2 (3 EXCLUDED)')
print(emp.iloc[0:3, [1, 2]])

<class 'pandas.Series'> ← single brackets → Series
<class 'pandas.DataFrame'> ← double brackets → DataFrame

loc[0:3, ["Name","Age"]] → labels 0..3 INCLUDED
    Name  Age
0    Ali   25
1   Sara   32
2  Ahmed   41
3   Hina   29 

iloc[0:3, [1, 2]] → positions 0..2 (3 EXCLUDED)
    Name  Age
0    Ali   25
1   Sara   32
2  Ahmed   41


In [3]:
# The difference becomes clear with a non-default index
by_name = emp.set_index('Name')                      # index labels are now names
print('loc["Sara", "Salary"]  →', by_name.loc['Sara', 'Salary'])      # by label
print('iloc[1, 4]             →', by_name.iloc[1, 4])                 # 2nd row, 5th column (Salary)
print(by_name.loc['Sara':'Omar', ['City', 'Salary']])                # label slice, end INCLUDED

sorted_emp = emp.sort_values('Salary', ascending=False)               # index is now shuffled
print('\nAfter sorting by salary: index =', sorted_emp.index.tolist())
print('sorted_emp.loc[0, "Name"]  →', sorted_emp.loc[0, 'Name'], '  (label 0 = Ali, wherever he is)')
print('sorted_emp.iloc[0]["Name"] →', sorted_emp.iloc[0]['Name'], ' (first ROW = highest salary)')

loc["Sara", "Salary"]  → 60000
iloc[1, 4]             → 60000
            City  Salary
Name                    
Sara     Karachi   60000
Ahmed     Lahore  120000
Hina   Islamabad   90000
Omar     Karachi   98000

After sorting by salary: index = [6, 2, 7, 4, 3, 0, 9, 1, 5, 8]
sorted_emp.loc[0, "Name"]  → Ali   (label 0 = Ali, wherever he is)
sorted_emp.iloc[0]["Name"] → Bilal  (first ROW = highest salary)


### Explanation
After sorting, label `0` still belongs to Ali, but position `0` is now the highest earner (Bilal). Mixing these up is a classic bug.

### When not
Don't use chained selection to **assign** values (`df['col'][0] = x`) — see Section 3.

### Common mistakes
- `df['col1', 'col2']` (missing inner brackets) → `KeyError`.
- Expecting `iloc[0:3]` to include position 3 (it doesn't) or `loc[0:3]` to exclude label 3 (it doesn't).
- Using `loc` with positions after the index changed.

---
<a id="sec-2"></a>
# 2. Filtering Rows

### What & why
Keeping only the rows that meet a **condition** ("employees in Lahore older than 30").

### How it works
A condition like `df['Age'] > 30` produces a **boolean Series** (True/False per row). Passing it inside `df[...]` keeps the True rows.

| Operator | Meaning | Example |
|---|---|---|
| `&` | AND | `(df['Age'] > 30) & (df['City'] == 'Lahore')` |
| `\|` | OR | `(df['City'] == 'Lahore') \| (df['City'] == 'Karachi')` |
| `~` | NOT | `~(df['Department'] == 'IT')` |
| `.isin([...])` | value in a list | `df['City'].isin(['Lahore', 'Karachi'])` |
| `.between(a, b)` | a ≤ value ≤ b (inclusive by default) | `df['Age'].between(25, 35)` |
| `.query('...')` | condition as a string | `df.query('Age > 30 and City == "Lahore"')` |

### ⚠️ Why parentheses are required
`&` and `|` have **higher priority** than `>` and `==` in Python. So
`df['Age'] > 30 & df['City'] == 'Lahore'` is read as `df['Age'] > (30 & df['City']) == 'Lahore'` → error or nonsense. Always wrap each condition in `( )`.

Also: use `&`, `|`, `~` — **not** Python's `and`, `or`, `not` (those don't work element-wise on Series → `ValueError: The truth value of a Series is ambiguous`).

In [4]:
print(emp[emp['Age'] > 30][['Name', 'Age']], '\n')
print(emp[emp['City'] == 'Lahore'][['Name', 'City']], '\n')

lahore_senior = emp[(emp['City'] == 'Lahore') & (emp['Age'] > 30)]           # AND
lahore_or_karachi = emp[(emp['City'] == 'Lahore') | (emp['City'] == 'Karachi')]  # OR
not_it = emp[~(emp['Department'] == 'IT')]                                     # NOT
print('AND :', lahore_senior['Name'].tolist())
print('OR  :', lahore_or_karachi['Name'].tolist())
print('NOT :', not_it['Name'].tolist())

     Name  Age
1    Sara   32
2   Ahmed   41
4    Omar   35
6   Bilal   45
7  Ayesha   38
9  Fatima   31 

    Name    City
0    Ali  Lahore
2  Ahmed  Lahore
5   Zara  Lahore
8  Usman  Lahore 

AND : ['Ahmed']
OR  : ['Ali', 'Sara', 'Ahmed', 'Omar', 'Zara', 'Ayesha', 'Usman']
NOT : ['Sara', 'Hina', 'Zara', 'Bilal', 'Usman', 'Fatima']


In [5]:
print('isin    :', emp[emp['Department'].isin(['IT', 'Finance'])]['Name'].tolist())
print('between :', emp[emp['Age'].between(25, 35)]['Name'].tolist())
min_salary = 90000
print('query   :', emp.query('Salary >= @min_salary and City != "Karachi"')['Name'].tolist())   # @ uses a Python variable

try:
    emp[emp['Age'] > 30 & emp['City'] == 'Lahore']          # ❌ missing parentheses
except Exception as e:
    print('\nWithout parentheses →', type(e).__name__)

isin    : ['Ali', 'Ahmed', 'Hina', 'Omar', 'Bilal', 'Ayesha']
between : ['Ali', 'Sara', 'Hina', 'Omar', 'Zara', 'Fatima']
query   : ['Ahmed', 'Hina', 'Bilal']

Without parentheses → TypeError


### When not
- `query()` is great for readability, but column names with spaces need backticks (`` `Monthly Salary` > 5 ``), and it's harder to build dynamically.
- For many OR conditions on one column, use `isin()` instead of chaining `|`.

### Common mistakes
Missing parentheses · using `and`/`or` · forgetting that `between` includes both ends · filtering creates a **new** DataFrame (assign it to a variable).

---
<a id="sec-3"></a>
# 3. Adding and Modifying Columns

### Syntax
```python
df['new'] = expression                               # add or overwrite a column
df.loc[condition, 'col'] = value                     # modify only some rows (safe way)
df['flag'] = np.where(condition, 'yes', 'no')        # if/else
df['level'] = np.select([cond1, cond2], ['A', 'B'], default='C')   # if / elif / else
df = df.assign(new=lambda d: d['a'] * 2)             # returns a NEW DataFrame (chainable)
```

| Tool | Use when |
|---|---|
| Arithmetic (`df['a'] * df['b']`) | Formula columns |
| `np.where(cond, x, y)` | Two outcomes |
| `np.select(conds, choices, default)` | Several ordered conditions (first True wins) |
| `assign()` | Method chaining; creating several columns in one expression |

### ⚠️ Copy-on-Write (pandas 3): no chained assignment
`df['Salary'][0] = 1` or `df[df['Age'] > 30]['Salary'] = 1` **never** modify `df` in pandas 3 (you get a `ChainedAssignmentError` warning). Use **one** `.loc[...] = ...` instead.

In [6]:
e = emp.copy()                                                     # work on a copy so `emp` stays unchanged
e['Annual_Salary'] = e['Salary'] * 12                                 # arithmetic
e['Salary_per_Exp_Year'] = (e['Salary'] / e['Experience']).round(0)
e['Senior'] = np.where(e['Experience'] >= 10, 'Yes', 'No')           # two outcomes
e['Band'] = np.select(
    [e['Salary'] >= 110000, e['Salary'] >= 80000],                    # checked in order
    ['High', 'Medium'],
    default='Low')
e['Salary'] = e['Salary'].astype(float)                              # int → float first (a 10% raise creates decimals)
e.loc[e['Department'] == 'Sales', 'Salary'] = e['Salary'] * 1.10     # modify some rows safely with ONE .loc
e[['Name', 'Department', 'Salary', 'Annual_Salary', 'Senior', 'Band']]

,Name,Department,Salary,Annual_Salary,Senior,Band
0,Ali,IT,85000.0,1020000,No,Medium
1,Sara,HR,60000.0,720000,No,Low
2,Ahmed,IT,120000.0,1440000,Yes,High
3,Hina,Finance,90000.0,1080000,No,Medium
4,Omar,IT,98000.0,1176000,No,Medium
5,Zara,HR,55000.0,660000,No,Low
6,Bilal,Finance,130000.0,1560000,Yes,High
7,Ayesha,IT,110000.0,1320000,Yes,High
8,Usman,Sales,49500.0,540000,No,Low
9,Fatima,Sales,77000.0,840000,No,Low


In [7]:
# assign(): build new columns in a chain without touching the original
summary = (emp
           .assign(Bonus=lambda d: d['Salary'] * 0.1,
                   Total=lambda d: d['Salary'] + d['Bonus'])        # can use Bonus created just before
           .loc[:, ['Name', 'Salary', 'Bonus', 'Total']])
print(summary.head(3))
print('\nIs there a Bonus column in the original emp?', 'Bonus' in emp.columns, '← assign() did not modify it')

    Name  Salary    Bonus     Total
0    Ali   85000   8500.0   93500.0
1   Sara   60000   6000.0   66000.0
2  Ahmed  120000  12000.0  132000.0

Is there a Bonus column in the original emp? False ← assign() did not modify it


### Common mistakes
- `np.select` conditions in the wrong order (the first True wins — put the strictest first).
- Chained assignment (silently does nothing in pandas 3).
- Writing floats into an **integer** column with `.loc` → pandas 3 raises `TypeError: Invalid value ... for dtype 'int64'`. Convert first (`astype(float)`) or round the new values to integers. (Writing `NaN` into an int column is allowed; the column becomes float.)
- Integer division surprises / dividing by zero → `inf` (check with `replace([np.inf, -np.inf], np.nan)`).

---
<a id="sec-4"></a>
# 4. Sorting

### Syntax
```python
df.sort_values('col', ascending=False)                         # by values
df.sort_values(['col1', 'col2'], ascending=[True, False])      # by several columns
df.sort_values('col', na_position='first')                     # where NaNs go ('last' by default)
df.sort_index()                                                # by the index labels
```

### Why / when
Rankings, top-N lists, time order (before rolling/shift/diff!), readable reports.

### When not
Sorting huge DataFrames just to get the top 5 → `nlargest(5, 'col')` / `nsmallest` is faster.

In [8]:
print(emp.sort_values('Salary', ascending=False)[['Name', 'Salary']].head(3), '\n')
print(emp.sort_values(['City', 'Salary'], ascending=[True, False])[['City', 'Name', 'Salary']], '\n')

with_nan = emp[['Name', 'Salary']].copy()
with_nan.loc[[2, 5], 'Salary'] = np.nan
print(with_nan.sort_values('Salary', na_position='first').head(4), '\n')
print('sort_index after sorting by salary:', emp.sort_values('Salary').sort_index().index.tolist())
print('nlargest(3):', emp.nlargest(3, 'Salary')['Name'].tolist())

     Name  Salary
6   Bilal  130000
2   Ahmed  120000
7  Ayesha  110000 

        City    Name  Salary
6  Islamabad   Bilal  130000
3  Islamabad    Hina   90000
9  Islamabad  Fatima   70000
7    Karachi  Ayesha  110000
4    Karachi    Omar   98000
1    Karachi    Sara   60000
2     Lahore   Ahmed  120000
0     Lahore     Ali   85000
5     Lahore    Zara   55000
8     Lahore   Usman   45000 

    Name   Salary
2  Ahmed      NaN
5   Zara      NaN
8  Usman  45000.0
1   Sara  60000.0 

sort_index after sorting by salary: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
nlargest(3): ['Bilal', 'Ahmed', 'Ayesha']


### Explanation
Multiple columns sort by the first column, then break ties by the second. `ascending` can be a list (one per column). Missing values go to the end unless `na_position='first'`.

### Common mistakes
Forgetting to assign the result (`df.sort_values(...)` returns a new DataFrame) · forgetting to `reset_index(drop=True)` if you later use positions.

---
<a id="sec-5"></a>
# 5. Renaming

```python
df.rename(columns={'old': 'new'})          # columns
df.rename(index={0: 'first'})              # index labels
df.columns = [...]                         # replace ALL names at once (must match length)
df.rename(columns=str.lower)               # apply a function to every name
```
**Why:** clean, consistent, meaningful names (no spaces, consistent case) make code simpler and prevent bugs.

In [9]:
renamed = emp.rename(columns={'Emp_ID': 'Employee_ID', 'Salary': 'Monthly_Salary'})
print(renamed.columns.tolist())
print(emp.rename(columns=str.lower).columns.tolist())
print(emp.set_index('Name').rename(index={'Ali': 'Ali Khan'}).index[:3].tolist())

messy = pd.DataFrame(columns=[' Order ID', 'Customer Name ', 'UNIT PRICE'])
messy.columns = messy.columns.str.strip().str.lower().str.replace(' ', '_')   # clean all names at once
print(messy.columns.tolist())

['Employee_ID', 'Name', 'Age', 'City', 'Department', 'Monthly_Salary', 'Experience']
['emp_id', 'name', 'age', 'city', 'department', 'salary', 'experience']
['Ali Khan', 'Sara', 'Ahmed']
['order_id', 'customer_name', 'unit_price']


### Common mistakes
Misspelling an old name in `rename` (no error — the column is just not renamed) · assigning `df.columns = [...]` with the wrong number of names.

---
<a id="sec-6"></a>
# 6. Dropping Data

```python
df.drop(columns=['col1', 'col2'])          # drop columns (clearest form)
df.drop(index=[0, 3])                      # drop rows by label
df.drop('col', axis=1)                     # axis=1 → columns, axis=0 → rows
df.dropna()                                # drop rows with ANY missing value
df.dropna(subset=['Salary'])               # only consider some columns
df.dropna(thresh=5)                        # keep rows with at least 5 non-missing values
df.dropna(axis=1)                          # drop COLUMNS containing missing values
```

### `axis` in one sentence
`axis=0` = **rows** (move down), `axis=1` = **columns** (move across). Using `columns=` / `index=` avoids the confusion.

### Why `inplace=True` is often avoided
- It **doesn't** make things faster or save memory in practice; pandas usually creates a copy internally anyway.
- It returns `None`, so it breaks **method chaining** and causes bugs like `df = df.drop(..., inplace=True)` → `df` becomes `None`!
- The pandas developers discourage it. Prefer `df = df.drop(...)`.

In [10]:
print(emp.drop(columns=['Emp_ID', 'Experience']).columns.tolist())
print(emp.drop(index=[0, 1]).index.tolist())

m = emp[['Name', 'Age', 'Salary']].copy()
m.loc[[1, 4], 'Salary'] = np.nan; m.loc[4, 'Age'] = np.nan
print('\ndropna()            →', len(m.dropna()), 'rows')
print('dropna(subset=Age)  →', len(m.dropna(subset=['Age'])), 'rows')
print('dropna(axis=1)      →', m.dropna(axis=1).columns.tolist())

result = m.drop(columns='Age', inplace=True)
print('\nWhat inplace=True returns:', result, '← so never write m = m.drop(..., inplace=True)')

['Name', 'Age', 'City', 'Department', 'Salary']
[2, 3, 4, 5, 6, 7, 8, 9]

dropna()            → 8 rows
dropna(subset=Age)  → 9 rows
dropna(axis=1)      → ['Name']

What inplace=True returns: None ← so never write m = m.drop(..., inplace=True)


### Common mistakes
`dropna()` without `subset` dropping far more rows than intended · the `inplace=True` + assignment trap · dropping rows before understanding **why** values are missing (Topic 3).

---
<a id="sec-7"></a>
# 7. Index Manipulation

```python
df.set_index('col')                 # use a column as the index (row labels)
df.reset_index()                    # move the index back into a column
df.reset_index(drop=True)           # discard the old index, get 0..n-1
```

### Why indexes matter
- **Fast, readable lookups** by label: `df.loc['Sara']`.
- **Automatic alignment**: arithmetic between two Series lines up by index labels.
- Many operations produce a meaningful index: `groupby`, `pivot_table`, `resample` (time index).

### When not
Don't hide important columns in the index when you'll need them as normal columns (merging, plotting, exporting) — `reset_index()` brings them back.

In [11]:
by_id = emp.set_index('Emp_ID')
print(by_id.loc[105, ['Name', 'Salary']], '\n')

bonus = pd.Series({103: 5000, 101: 2000, 999: 1000})             # different order, one unknown ID
print('Automatic alignment by index (NaN where no match):')
print((by_id['Salary'] + bonus).head(4), '\n')

city_avg = emp.groupby('City')['Salary'].mean()                   # City becomes the index
print(city_avg, '\n')
print(city_avg.reset_index())                                     # back to a normal column

filtered = emp[emp['Age'] > 35]
print('\nIndex after filtering:', filtered.index.tolist(), '→ reset_index(drop=True):', filtered.reset_index(drop=True).index.tolist())

Name       Omar
Salary    98000
Name: 105, dtype: object 

Automatic alignment by index (NaN where no match):
Emp_ID
101     87000.0
102         NaN
103    125000.0
104         NaN
dtype: float64 

City
Islamabad    96666.666667
Karachi      89333.333333
Lahore       76250.000000
Name: Salary, dtype: float64 

        City        Salary
0  Islamabad  96666.666667
1    Karachi  89333.333333
2     Lahore  76250.000000

Index after filtering: [2, 6, 7] → reset_index(drop=True): [0, 1, 2]


### Common mistakes
Forgetting `drop=True` (adds an unwanted `index` column) · duplicate index labels making `loc` return several rows · unexpected NaNs caused by index alignment.

---
<a id="sec-8"></a>
# 8. Applying Functions — `apply`, `map`, `DataFrame.map` (formerly `applymap`)

| Method | Works on | Function receives | Typical use |
|---|---|---|---|
| `Series.map(func or dict)` | One column | **Each value** | Transform/look up values: `s.map({'M': 'Male'})` |
| `Series.apply(func)` | One column | Each value | Like `map` for functions |
| `DataFrame.apply(func, axis=0)` | DataFrame | **Each column** (a Series) | Column-wise summaries |
| `DataFrame.apply(func, axis=1)` | DataFrame | **Each row** (a Series) | Logic that needs several columns (slow!) |
| `DataFrame.map(func)` | DataFrame | **Each cell** | Element-wise formatting (was `applymap`) |

⚠️ **`applymap()` was deprecated in pandas 2.1 and removed in pandas 3.0** → use `DataFrame.map()`.

### Modern alternative: vectorized operations
`apply` with a Python function runs a **Python loop** internally. Built-in vectorized operations (`df['a'] * df['b']`, `np.where`, `.str`, `.dt` methods) run in optimized C code and are often **10–100× faster**. Use `apply` only when no vectorized option exists.

In [12]:
print(emp['Department'].map({'IT': 'Tech', 'HR': 'People', 'Finance': 'Money', 'Sales': 'Revenue'}).head(4).tolist())
print(emp['Name'].map(len).head(4).tolist(), '← map with a function')
print(emp['Salary'].apply(lambda s: 'high' if s > 90000 else 'normal').head(4).tolist())

print('\napply on columns (axis=0):')
print(emp[['Age', 'Salary', 'Experience']].apply(lambda col: col.max() - col.min()))

print('\napply on rows (axis=1):')
emp_rows = emp.apply(lambda r: f"{r['Name']} ({r['City']})", axis=1)
print(emp_rows.head(3).tolist())

print('\nDataFrame.map (element-wise, replaces applymap):')
print(emp[['Salary', 'Age']].head(3).map(lambda v: f'{v:,}'))

['Tech', 'People', 'Tech', 'Money']
[3, 4, 5, 4] ← map with a function
['normal', 'normal', 'high', 'normal']

apply on columns (axis=0):
Age              22
Salary        85000
Experience       19
dtype: int64

apply on rows (axis=1):
['Ali (Lahore)', 'Sara (Karachi)', 'Ahmed (Lahore)']

DataFrame.map (element-wise, replaces applymap):
    Salary Age
0   85,000  25
1   60,000  32
2  120,000  41


In [13]:
import time
big = pd.DataFrame({'price': np.random.default_rng(0).uniform(10, 1000, 300_000),
                    'qty': np.random.default_rng(1).integers(1, 10, 300_000)})

t0 = time.perf_counter(); r1 = big.apply(lambda r: r['price'] * r['qty'], axis=1); t_apply = time.perf_counter() - t0
t0 = time.perf_counter(); r2 = big['price'] * big['qty'];                             t_vec = time.perf_counter() - t0
print(f'apply(axis=1): {t_apply:.3f} s | vectorized: {t_vec:.4f} s | same result: {np.allclose(r1, r2)}')
print(f'→ vectorized is about {t_apply / t_vec:,.0f}× faster')

apply(axis=1): 1.102 s | vectorized: 0.0008 s | same result: True
→ vectorized is about 1,408× faster


### Common mistakes
Using `apply(axis=1)` for simple arithmetic · `map` with a dict that misses some keys → those values become **NaN** · calling the removed `applymap` in pandas 3.

---
<a id="sec-9"></a>
# 9. GroupBy (deep dive)

### What
**Split → Apply → Combine**:
```text
Split the rows into groups (e.g. by City) → apply a function to each group (e.g. mean salary) → combine the results into one table
```

### Syntax
```python
df.groupby('City')['Salary'].mean()                        # one column, one statistic
df.groupby('City')[['Salary', 'Age']].agg(['mean', 'max'])  # several statistics
df.groupby(['City', 'Department'])['Salary'].sum()          # several grouping keys → MultiIndex
df.groupby('City').agg(                                     # NAMED aggregation (recommended)
    avg_salary=('Salary', 'mean'),
    max_salary=('Salary', 'max'),
    employees=('Salary', 'count'),
)
```

Common functions: `sum, mean, median, count, size, min, max, std, nunique, first, last`.
- `count` = non-missing values; `size` = all rows (including NaN).

### Why / when
Any question with "**per** …" or "**by** …": average salary **per** city, total sales **by** month, number of orders **per** customer.

### When not
When you need the group statistic **on every row** → `transform` (Section 10). For a 2-D summary table → `pivot_table` (Section 12).

In [14]:
print('Mean salary by city:'); print(emp.groupby('City')['Salary'].mean().round(0), '\n')
for f in ['sum', 'median', 'count', 'min', 'max', 'std']:
    print(f'{f:<7}', emp.groupby('City')['Salary'].agg(f).round(0).to_dict())

Mean salary by city:
City
Islamabad    96667.0
Karachi      89333.0
Lahore       76250.0
Name: Salary, dtype: float64 

sum     {'Islamabad': 290000, 'Karachi': 268000, 'Lahore': 305000}
median  {'Islamabad': 90000.0, 'Karachi': 98000.0, 'Lahore': 70000.0}
count   {'Islamabad': 3, 'Karachi': 3, 'Lahore': 4}
min     {'Islamabad': 70000, 'Karachi': 60000, 'Lahore': 45000}
max     {'Islamabad': 130000, 'Karachi': 110000, 'Lahore': 120000}
std     {'Islamabad': 30551.0, 'Karachi': 26102.0, 'Lahore': 33758.0}


In [15]:
print('Multiple aggregations:')
print(emp.groupby('Department')[['Salary', 'Experience']].agg(['mean', 'max']).round(1), '\n')

print('Named aggregation (clean column names):')
city_summary = emp.groupby('City').agg(
    avg_salary=('Salary', 'mean'),
    max_salary=('Salary', 'max'),
    employees=('Salary', 'count'),
    avg_age=('Age', 'mean'),
).round(1)
print(city_summary, '\n')

print('Two grouping keys:')
print(emp.groupby(['City', 'Department'])['Salary'].sum().reset_index())

Multiple aggregations:
              Salary         Experience    
                mean     max       mean max
Department                                 
Finance     110000.0  130000       12.5  20
HR           57500.0   60000        4.5   6
IT          103250.0  120000        9.5  15
Sales        57500.0   70000        4.0   7 

Named aggregation (clean column names):
           avg_salary  max_salary  employees  avg_age
City                                                 
Islamabad     96666.7      130000          3     35.0
Karachi       89333.3      110000          3     35.0
Lahore        76250.0      120000          4     29.0 

Two grouping keys:
        City Department  Salary
0  Islamabad    Finance  220000
1  Islamabad      Sales   70000
2    Karachi         HR   60000
3    Karachi         IT  208000
4     Lahore         HR   55000
5     Lahore         IT  205000
6     Lahore      Sales   45000


### Expected output
Named aggregation gives a tidy table with **one row per city** and readable column names (`avg_salary`, `max_salary`, …).

### Common mistakes
- Forgetting `reset_index()` when you need the group keys as columns (or use `as_index=False`).
- `count` vs `size` confusion.
- Grouping by a column with NaN: by default NaN groups are **dropped** (`dropna=False` keeps them).
- Multi-level column names from `.agg(['mean', 'max'])` that are awkward to use → prefer named aggregation.

---
<a id="sec-10"></a>
# 10. `groupby().transform()`

### What
Computes a group statistic but returns a result **the same length as the original DataFrame** — each row gets the value of **its own group**.

| | `agg()` | `transform()` |
|---|---|---|
| Output rows | **One per group** | **Same as original** (one per row) |
| Use for | Summary tables | New columns: group mean on each row, % of group, rank within group, standardise within group, fill NaN with group mean |

### Syntax
```python
df['City_Avg'] = df.groupby('City')['Salary'].transform('mean')
```

In [16]:
t = emp[['Name', 'City', 'Salary']].copy()
print('agg → one row per city:', emp.groupby('City')['Salary'].agg('mean').shape)
t['City_Avg'] = t.groupby('City')['Salary'].transform('mean').round(0)               # repeated for each row
t['Diff_vs_City'] = t['Salary'] - t['City_Avg']
t['Share_of_City_Total_%'] = (t['Salary'] / t.groupby('City')['Salary'].transform('sum') * 100).round(1)
t['Rank_in_City'] = t.groupby('City')['Salary'].rank(ascending=False).astype(int)
t.sort_values(['City', 'Rank_in_City'])

agg → one row per city: (3,)


,Name,City,Salary,City_Avg,Diff_vs_City,Share_of_City_Total_%,Rank_in_City
6,Bilal,Islamabad,130000,96667.0,33333.0,44.8,1
3,Hina,Islamabad,90000,96667.0,-6667.0,31.0,2
9,Fatima,Islamabad,70000,96667.0,-26667.0,24.1,3
7,Ayesha,Karachi,110000,89333.0,20667.0,41.0,1
4,Omar,Karachi,98000,89333.0,8667.0,36.6,2
1,Sara,Karachi,60000,89333.0,-29333.0,22.4,3
2,Ahmed,Lahore,120000,76250.0,43750.0,39.3,1
0,Ali,Lahore,85000,76250.0,8750.0,27.9,2
5,Zara,Lahore,55000,76250.0,-21250.0,18.0,3
8,Usman,Lahore,45000,76250.0,-31250.0,14.8,4


In [17]:
# Practical: fill missing salaries with the average of the same department
s_missing = emp[['Name', 'Department', 'Salary']].copy()
s_missing.loc[[0, 6], 'Salary'] = np.nan
s_missing['Salary_Filled'] = s_missing['Salary'].fillna(s_missing.groupby('Department')['Salary'].transform('mean'))
s_missing

,Name,Department,Salary,Salary_Filled
0,Ali,IT,NaN,109333.333333
1,Sara,HR,60000.0,60000.000000
2,Ahmed,IT,120000.0,120000.000000
3,Hina,Finance,90000.0,90000.000000
4,Omar,IT,98000.0,98000.000000
5,Zara,HR,55000.0,55000.000000
6,Bilal,Finance,NaN,90000.000000
7,Ayesha,IT,110000.0,110000.000000
8,Usman,Sales,45000.0,45000.000000
9,Fatima,Sales,70000.0,70000.000000


### Common mistakes
Using `agg` and then trying to assign it as a column (lengths don't match / misaligned) · forgetting `transform` keeps the **original index** (that's why assignment works).

---
<a id="sec-11"></a>
# 11. `groupby().filter()`

### What
Keeps or drops **whole groups** based on a condition computed on the group. The result has the **original rows** (not a summary).

```python
df.groupby('City').filter(lambda g: len(g) >= 3)           # keep cities with at least 3 employees
df.groupby('Department').filter(lambda g: g['Salary'].mean() > 80000)
```

### When
"Only analyse customers with ≥ 5 orders", "only products sold in ≥ 3 cities", "remove tiny groups before computing group statistics".

### When not
Filtering individual rows → use a normal boolean filter. On large data with many groups, `filter` with a Python function can be slow → a `transform`-based mask is faster:
`df[df.groupby('City')['City'].transform('size') >= 3]`.

In [18]:
big_cities = emp.groupby('City').filter(lambda g: len(g) >= 4)
print('Cities with ≥ 4 employees:', big_cities['City'].unique().tolist(), '| rows kept:', len(big_cities))

rich_depts = emp.groupby('Department').filter(lambda g: g['Salary'].mean() > 80000)
print('Departments with mean salary > 80k:', rich_depts['Department'].unique().tolist())

fast = emp[emp.groupby('City')['City'].transform('size') >= 4]
print('Same result with transform mask:', fast.equals(big_cities))

Cities with ≥ 4 employees: ['Lahore'] | rows kept: 4
Departments with mean salary > 80k: ['IT', 'Finance']
Same result with transform mask: True


---
<a id="sec-12"></a>
# 12. Pivot Tables — `pd.pivot_table()`

### What
A 2-D summary table (like Excel's PivotTable): **rows** from one column, **columns** from another, **cells** = an aggregate of a third column.

```python
pd.pivot_table(df,
               index='City',            # rows
               columns='Department',    # columns
               values='Salary',         # what to summarise
               aggfunc='mean',          # how: 'mean', 'sum', 'count', ['mean', 'max'], {'Salary': 'mean', 'Age': 'max'}
               fill_value=0,            # replace empty cells
               margins=True,            # add an 'All' row/column with totals
               margins_name='Total')
```

### When / when not
✅ Two-dimensional summaries for reports and heatmaps. ❌ Just counts of two categories → `crosstab` is shorter. ❌ One grouping key → `groupby` is simpler.

In [19]:
pt = pd.pivot_table(emp, index='City', columns='Department', values='Salary', aggfunc='mean', fill_value=0,
                    margins=True, margins_name='Total').round(0)
print(pt, '\n')
print(pd.pivot_table(emp, index='City', values=['Salary', 'Age'], aggfunc={'Salary': 'sum', 'Age': 'mean'}).round(1))

Department   Finance       HR        IT    Sales    Total
City                                                     
Islamabad   110000.0      0.0       0.0  70000.0  96667.0
Karachi          0.0  60000.0  104000.0      0.0  89333.0
Lahore           0.0  55000.0  102500.0  45000.0  76250.0
Total       110000.0  57500.0  103250.0  57500.0  86300.0 

            Age  Salary
City                   
Islamabad  35.0  290000
Karachi    35.0  268000
Lahore     29.0  305000


### Explanation
Each cell is the average salary of that city–department combination; `0` means no employees there (from `fill_value=0`). The `Total` row/column is computed from the **raw data**, not by averaging the averages.

### Common mistakes
`fill_value=0` for a **mean** can be misleading (0 ≠ "no data" — consider leaving NaN) · forgetting `values=` (pandas aggregates every numeric column) · confusing `pivot` (no aggregation, fails on duplicates — Section 21) with `pivot_table`.

---
<a id="sec-13"></a>
# 13. Crosstab — `pd.crosstab()`

### What
A frequency table: **how many rows** fall into each combination of two (or more) categorical variables.

```python
pd.crosstab(df['City'], df['Department'])                                  # counts
pd.crosstab(df['City'], df['Department'], normalize='index')               # row percentages
pd.crosstab(df['City'], df['Department'], values=df['Salary'], aggfunc='mean')   # like a pivot table
pd.crosstab(df['City'], df['Department'], margins=True)                   # totals
```

### When useful
Checking relationships between categories (before a chi-square test, Topic 14), class balance by segment, survey tables. `normalize='index'` answers "within each city, what share works in each department?".

In [20]:
print(pd.crosstab(emp['City'], emp['Department'], margins=True), '\n')
print((pd.crosstab(emp['City'], emp['Department'], normalize='index') * 100).round(0), '\n')
print(pd.crosstab(emp['City'], emp['Department'], values=emp['Salary'], aggfunc='mean').round(0))

Department  Finance  HR  IT  Sales  All
City                                   
Islamabad         2   0   0      1    3
Karachi           0   1   2      0    3
Lahore            0   1   2      1    4
All               2   2   4      2   10 

Department  Finance    HR    IT  Sales
City                                  
Islamabad      67.0   0.0   0.0   33.0
Karachi         0.0  33.0  67.0    0.0
Lahore          0.0  25.0  50.0   25.0 

Department   Finance       HR        IT    Sales
City                                            
Islamabad   110000.0      NaN       NaN  70000.0
Karachi          NaN  60000.0  104000.0      NaN
Lahore           NaN  55000.0  102500.0  45000.0


### Common mistakes
Passing column **names** instead of Series (`pd.crosstab('City', 'Department')` — it needs `df['City']`) · mixing `normalize='index'` (rows sum to 1) and `'columns'` (columns sum to 1).

---
<a id="sec-14"></a>
# 14. String Manipulation — the `.str` accessor

| Method | What it does | Example result |
|---|---|---|
| `.str.lower()` / `.str.upper()` / `.str.title()` | Change case | `'lahore'`, `'LAHORE'`, `'Lahore'` |
| `.str.strip()` | Remove spaces at both ends | `'  Ali '` → `'Ali'` |
| `.str.contains('x', case=False)` | True if the text contains a pattern | Filter emails from gmail |
| `.str.startswith('+92')` / `.str.endswith('.pk')` | Check beginning / end | Phone/domain checks |
| `.str.replace('-', '', regex=False)` | Replace text | Clean phone numbers |
| `.str.split('@')` / `.str.split('@', expand=True)` | Split into a list / into columns | User & domain |
| `.str.len()` | Length of each text | Detect short/invalid codes |
| `.str.extract(r'(\d+)')` | Pull out a **regex** group | Numbers inside text |

### Why
Real text data is messy: extra spaces, mixed case, inconsistent formats. Vectorized `.str` methods clean an entire column in one line (and handle missing values gracefully).

In [21]:
customers = pd.DataFrame({
    'Name':  ['  ali khan', 'SARA AHMED ', 'Omar farooq', 'hina  Malik', None],
    'Email': ['ali.khan@gmail.com', 'sara@company.pk', 'OMAR@Gmail.com', 'hina@yahoo.com', 'test@company.pk'],
    'Phone': ['+92-300-1234567', '0321 7654321', '+92 333 1112223', '0300-9998887', '+92-345-0001112'],
    'Product_Code': ['ELEC-1001', 'FASH-2040', 'ELEC-1500', 'BOOK-3002', 'FASH-2999'],
    'Address': ['House 12, Street 5, Lahore', 'Flat 3B, Clifton, Karachi', 'Plot 44, F-7, Islamabad', 'House 9, Gulberg, Lahore', 'Street 2, Saddar, Karachi'],
})

c = customers.copy()
c['Name'] = c['Name'].str.strip().str.replace(r'\s+', ' ', regex=True).str.title()   # trim, collapse spaces, Title Case
c['Email'] = c['Email'].str.lower()
c['Is_Gmail'] = c['Email'].str.contains('gmail', case=False)
c['Is_PK_Domain'] = c['Email'].str.endswith('.pk')
c['Phone_Clean'] = c['Phone'].str.replace(r'[^\d]', '', regex=True).str.replace(r'^92', '0', regex=True)  # keep digits, +92 → 0
c['Phone_OK'] = c['Phone_Clean'].str.len() == 11
c['Intl_Format'] = c['Phone'].str.startswith('+92')
c[['User', 'Domain']] = c['Email'].str.split('@', expand=True)                       # split into two columns
c['Category_Code'] = c['Product_Code'].str.extract(r'^([A-Z]+)-')                    # letters before the dash
c['Code_Number'] = c['Product_Code'].str.extract(r'-(\d+)$').astype(int)             # digits after the dash
c['City'] = c['Address'].str.split(',').str[-1].str.strip()                         # last part of the address
c.drop(columns=['Phone', 'Address'])

,Name,Email,Product_Code,Is_Gmail,Is_PK_Domain,Phone_Clean,Phone_OK,Intl_Format,User,Domain,Category_Code,Code_Number,City
0,Ali Khan,ali.khan@gmail.com,ELEC-1001,True,False,03001234567,True,True,ali.khan,gmail.com,ELEC,1001,Lahore
1,Sara Ahmed,sara@company.pk,FASH-2040,False,True,03217654321,True,False,sara,company.pk,FASH,2040,Karachi
2,Omar Farooq,omar@gmail.com,ELEC-1500,True,False,03331112223,True,True,omar,gmail.com,ELEC,1500,Islamabad
3,Hina Malik,hina@yahoo.com,BOOK-3002,False,False,03009998887,True,False,hina,yahoo.com,BOOK,3002,Lahore
4,NaN,test@company.pk,FASH-2999,False,True,03450001112,True,True,test,company.pk,FASH,2999,Karachi


### Explanation
- Missing names stay missing (`None`/NaN) — `.str` methods don't crash on them.
- `regex=True` lets you use patterns: `\s+` = one or more spaces, `[^\d]` = anything that is not a digit, `^92` = "92" at the start.
- `str.extract` returns the part matched inside `( )`.

### Common mistakes
Forgetting `regex=False` when replacing special characters like `.` or `+` (they have special regex meanings) · `str.contains` on columns with NaN inside a filter → use `na=False` · calling `.str` on a numeric column (convert with `astype(str)` first).

---
<a id="sec-15"></a>
# 15. Date/Time Manipulation

```python
df['Date'] = pd.to_datetime(df['Date'], format='%d/%m/%Y')   # text → datetime (format avoids ambiguity)
df['Date'].dt.year / .dt.month / .dt.day / .dt.day_name() / .dt.hour / .dt.quarter / .dt.dayofweek
df[(df['Date'] >= '2025-03-01') & (df['Date'] < '2025-04-01')]    # date filtering
df.set_index('Date').loc['2025-03']                                # partial string selection (sorted index)
df.groupby(df['Date'].dt.to_period('M'))['Sales'].sum()            # monthly grouping
df.set_index('Date').resample('W')['Sales'].sum()                  # weekly totals
```

### Why
Dates stored as text can't be compared, sorted correctly (`'10/01'` < `'2/01'` as text!) or grouped by month. `datetime` dtype unlocks all time-based analysis.

### Common mistakes
- Ambiguous formats: `01/02/2025` is 1 February in Pakistan/UK but January 2 in the US → always pass `format=` (or `dayfirst=True`).
- Bad values → use `errors='coerce'` (they become `NaT`) and then inspect them.
- `resample` needs a datetime **index** (or `on='Date'`).

In [22]:
orders = pd.DataFrame({
    'Order_ID': range(1, 11),
    'Order_Time': ['05/01/2025 09:15', '17/01/2025 14:30', '02/02/2025 20:05', '14/02/2025 11:45', '28/02/2025 16:20',
                   '03/03/2025 08:10', '15/03/2025 19:55', '22/03/2025 13:00', 'not a date', '30/03/2025 22:40'],
    'Amount': [1200, 3400, 800, 5600, 2300, 4100, 900, 3000, 1500, 2700],
})
orders['Order_Time'] = pd.to_datetime(orders['Order_Time'], format='%d/%m/%Y %H:%M', errors='coerce')
print('Unparseable rows:', orders['Order_Time'].isna().sum())
ot = orders['Order_Time'].dt
orders = orders.assign(Year=ot.year.astype('Int64'), Month=ot.month.astype('Int64'), Day=ot.day.astype('Int64'),
                       Weekday=ot.day_name(), Hour=ot.hour.astype('Int64'),   # 'Int64' = integers that allow missing (<NA>)
                       Is_Weekend=ot.dayofweek >= 5, Evening=ot.hour >= 18)
orders

Unparseable rows: 1


,Order_ID,Order_Time,Amount,Year,Month,Day,Weekday,Hour,Is_Weekend,Evening
0,1,2025-01-05 09:15:00,1200,2025,1,5,Sunday,9,True,False
1,2,2025-01-17 14:30:00,3400,2025,1,17,Friday,14,False,False
2,3,2025-02-02 20:05:00,800,2025,2,2,Sunday,20,True,True
3,4,2025-02-14 11:45:00,5600,2025,2,14,Friday,11,False,False
4,5,2025-02-28 16:20:00,2300,2025,2,28,Friday,16,False,False
5,6,2025-03-03 08:10:00,4100,2025,3,3,Monday,8,False,False
6,7,2025-03-15 19:55:00,900,2025,3,15,Saturday,19,True,True
7,8,2025-03-22 13:00:00,3000,2025,3,22,Saturday,13,True,False
8,9,NaT,1500,<NA>,<NA>,<NA>,NaN,<NA>,False,False
9,10,2025-03-30 22:40:00,2700,2025,3,30,Sunday,22,True,True


In [23]:
valid = orders.dropna(subset=['Order_Time'])
print('March orders (date filter):', valid[(valid['Order_Time'] >= '2025-03-01') & (valid['Order_Time'] < '2025-04-01')]['Order_ID'].tolist())
print('\nMonthly revenue (to_period):'); print(valid.groupby(valid['Order_Time'].dt.to_period('M'))['Amount'].sum())
print('\nWeekly revenue (resample):'); print(valid.set_index('Order_Time').resample('W')['Amount'].sum().head(4))
print('\nDays between first and last order:', (valid['Order_Time'].max() - valid['Order_Time'].min()).days)

March orders (date filter): [6, 7, 8, 10]

Monthly revenue (to_period):
Order_Time
2025-01     4600
2025-02     8700
2025-03    10700
Freq: M, Name: Amount, dtype: int64

Weekly revenue (resample):
Order_Time
2025-01-05    1200
2025-01-12       0
2025-01-19    3400
2025-01-26       0
Freq: W-SUN, Name: Amount, dtype: int64

Days between first and last order: 84


---
<a id="sec-16"></a>
# 16. Duplicate Handling

```python
df.duplicated()                                   # True for rows that repeat an EARLIER row
df.duplicated(subset=['Email'], keep=False)       # mark ALL copies
df.drop_duplicates()                              # remove exact duplicate rows (keeps the first)
df.drop_duplicates(subset=['Email'], keep='last') # one row per email, keep the latest
```

| `keep=` | Marks as duplicate |
|---|---|
| `'first'` (default) | All copies except the first |
| `'last'` | All copies except the last |
| `False` | **All** copies (useful to inspect them) |

### When not
Don't blindly drop duplicates: two identical orders might be **real** (a customer bought the same item twice). Define what makes a row unique (e.g. `Order_ID`) and use `subset=`.

In [24]:
dups = pd.DataFrame({
    'Customer': ['Ali', 'Sara', 'Ali', 'Omar', 'Sara', 'Hina'],
    'Email': ['ali@x.com', 'sara@x.com', 'ali@x.com', 'omar@x.com', 'SARA@x.com', 'hina@x.com'],
    'Signup': pd.to_datetime(['2025-01-01', '2025-01-05', '2025-01-01', '2025-02-01', '2025-03-10', '2025-02-15']),
})
print('Exact duplicates:', dups.duplicated().sum())
print(dups[dups.duplicated(keep=False)], '\n')

dups['Email_Clean'] = dups['Email'].str.lower()                               # 'SARA@x.com' is the same person!
print('Duplicates by cleaned email:', dups.duplicated(subset=['Email_Clean']).sum())
latest = dups.sort_values('Signup').drop_duplicates(subset=['Email_Clean'], keep='last')
latest

Exact duplicates: 1
  Customer      Email     Signup
0      Ali  ali@x.com 2025-01-01
2      Ali  ali@x.com 2025-01-01 

Duplicates by cleaned email: 2


,Customer,Email,Signup,Email_Clean
2,Ali,ali@x.com,2025-01-01,ali@x.com
3,Omar,omar@x.com,2025-02-01,omar@x.com
5,Hina,hina@x.com,2025-02-15,hina@x.com
4,Sara,SARA@x.com,2025-03-10,sara@x.com


### Common mistakes
Checking duplicates **before** cleaning text (case/spaces hide duplicates) · forgetting to sort before `keep='last'` · removing legitimate repeated transactions.

---
<a id="sec-17"></a>
# 17. Missing Value Manipulation

| Method | What it does | Use when |
|---|---|---|
| `fillna(value)` | Replace NaN with a constant | Categorical "Unknown", 0 when missing truly means zero |
| `fillna(df['col'].median())` | Fill with a statistic | Numeric, roughly random missingness (fit on training data in ML!) |
| `ffill()` | Copy the **previous** value forward | Time series / sensor readings where the last known value still holds (stock price, status) |
| `bfill()` | Copy the **next** value backward | Gaps at the start of a series; when the next value is the best guess |
| `interpolate()` | Estimate between neighbours (linear by default; `method='time'` for dates) | Smooth numeric series (temperature, gradual trends) |

### When not
- `ffill`/`bfill`/`interpolate` on **unordered** data (sort first!) or across **different groups** (use `groupby().ffill()`).
- Interpolating categories or long gaps.
- Filling a target variable, or filling with statistics from the test set (leakage).

In [25]:
temps = pd.DataFrame({'Date': pd.date_range('2025-06-01', periods=8),
                      'Temp': [30.0, np.nan, 34.0, np.nan, np.nan, 40.0, np.nan, 38.0],
                      'Status': ['OK', None, None, 'Maintenance', None, 'OK', None, None]})
temps['fillna_mean'] = temps['Temp'].fillna(temps['Temp'].mean()).round(1)
temps['ffill'] = temps['Temp'].ffill()
temps['bfill'] = temps['Temp'].bfill()
temps['interpolate'] = temps['Temp'].interpolate()
temps['Status_ffill'] = temps['Status'].ffill()
temps

,Date,Temp,Status,fillna_mean,ffill,bfill,interpolate,Status_ffill
0,2025-06-01,30.0,OK,30.0,30.0,30.0,30.0,OK
1,2025-06-02,NaN,NaN,35.5,30.0,34.0,32.0,OK
2,2025-06-03,34.0,NaN,34.0,34.0,34.0,34.0,OK
3,2025-06-04,NaN,Maintenance,35.5,34.0,40.0,36.0,Maintenance
4,2025-06-05,NaN,NaN,35.5,34.0,40.0,38.0,Maintenance
5,2025-06-06,40.0,OK,40.0,40.0,40.0,40.0,OK
6,2025-06-07,NaN,NaN,35.5,40.0,38.0,39.0,OK
7,2025-06-08,38.0,NaN,38.0,38.0,38.0,38.0,OK


In [26]:
# Group-aware forward fill: never carry a value from one store into another
stock = pd.DataFrame({'Store': ['A', 'A', 'A', 'B', 'B', 'B'], 'Day': [1, 2, 3, 1, 2, 3], 'Stock': [50, np.nan, 45, np.nan, 30, np.nan]})
stock['Wrong_ffill'] = stock['Stock'].ffill()
stock['Group_ffill'] = stock.groupby('Store')['Stock'].ffill()
stock

,Store,Day,Stock,Wrong_ffill,Group_ffill
0,A,1,50.0,50.0,50.0
1,A,2,NaN,50.0,50.0
2,A,3,45.0,45.0,45.0
3,B,1,NaN,45.0,NaN
4,B,2,30.0,30.0,30.0
5,B,3,NaN,30.0,30.0


### Explanation
- Linear interpolation fills 34 → 40 with 36 and 38 (a straight line); `ffill` repeats 34; `bfill` uses 40.
- In the store example, the wrong `ffill` copies Store A's value (45) into Store B's first day; the group-aware version leaves it as NaN.

---
<a id="sec-18"></a>
# 18. Replace and Map

| | `replace()` | `map()` |
|---|---|---|
| Values **not** in the dictionary | **Kept unchanged** | Become **NaN** |
| Works on | Series or whole DataFrame | Series |
| Also supports | Regex, lists, nested dicts | Functions, Series lookups |
| Best for | Fixing a few specific values (typos, codes) | Complete recodings / lookups (every value has a mapping) |

```python
df['Gender'].replace({'M': 'Male', 'F': 'Female'})
df['Size'].map({'S': 1, 'M': 2, 'L': 3})          # ordinal encoding (Topic 6)
df.replace({'N/A': np.nan, '-': np.nan})          # standardise missing-value markers everywhere
```

In [27]:
survey = pd.DataFrame({'Gender': ['M', 'F', 'Female', 'm', 'Other'], 'Size': ['S', 'L', 'M', 'XL', 'M'],
                       'Rating': ['5', 'N/A', '4', '-', '3']})
s = survey.copy()
s['Gender_replace'] = s['Gender'].str.upper().replace({'M': 'Male', 'F': 'Female', 'FEMALE': 'Female'})   # unmatched kept
s['Gender_map'] = s['Gender'].str.upper().map({'M': 'Male', 'F': 'Female'})                             # unmatched → NaN
s['Size_Code'] = s['Size'].map({'S': 1, 'M': 2, 'L': 3, 'XL': 4})
s['Rating'] = s['Rating'].replace({'N/A': np.nan, '-': np.nan}).astype(float)
s

,Gender,Size,Rating,Gender_replace,Gender_map,Size_Code
0,M,S,5.0,Male,Male,1
1,F,L,NaN,Female,Female,3
2,Female,M,4.0,Female,NaN,2
3,m,XL,NaN,Male,Male,4
4,Other,M,3.0,OTHER,NaN,2


### Common mistakes
Using `map` with an incomplete dictionary (silent NaNs) · forgetting to clean case/spaces first · `replace` on a whole DataFrame changing values in columns you didn't intend.

---
<a id="sec-19"></a>
# 19. Binning — `pd.cut()` and `pd.qcut()`

| | `pd.cut` | `pd.qcut` |
|---|---|---|
| Bins by | **Value ranges** (fixed width, or your own edges) | **Quantiles** (equal number of rows per bin) |
| Example | Ages 0–18, 19–30, 31–50, 51+ | Salary quartiles (bottom 25% … top 25%) |
| Bin widths | Equal or chosen | Unequal |
| Rows per bin | Unequal | ≈ equal |

```python
pd.cut(df['Age'], bins=[0, 25, 35, 100], labels=['Young', 'Mid', 'Senior'])
pd.cut(df['Salary'], bins=3)                            # 3 equal-width bins
pd.qcut(df['Salary'], q=4, labels=['Q1', 'Q2', 'Q3', 'Q4'])
```
(See the KBinsDiscretizer notebook for the ML version.)

In [28]:
b = emp[['Name', 'Age', 'Salary']].copy()
b['Age_Group'] = pd.cut(b['Age'], bins=[0, 25, 35, 100], labels=['≤25', '26–35', '36+'])   # right-inclusive: (25, 35]
b['Salary_EqualWidth'] = pd.cut(b['Salary'], bins=3, labels=['Low', 'Mid', 'High'])
b['Salary_Quartile'] = pd.qcut(b['Salary'], q=4, labels=['Q1', 'Q2', 'Q3', 'Q4'])
print(b, '\n')
print('Rows per equal-width bin:', b['Salary_EqualWidth'].value_counts().sort_index().to_dict())
print('Rows per quartile       :', b['Salary_Quartile'].value_counts().sort_index().to_dict())

     Name  Age  Salary Age_Group Salary_EqualWidth Salary_Quartile
0     Ali   25   85000       ≤25               Mid              Q2
1    Sara   32   60000     26–35               Low              Q1
2   Ahmed   41  120000       36+              High              Q4
3    Hina   29   90000     26–35               Mid              Q3
4    Omar   35   98000     26–35               Mid              Q3
5    Zara   27   55000     26–35               Low              Q1
6   Bilal   45  130000       36+              High              Q4
7  Ayesha   38  110000       36+              High              Q4
8   Usman   23   45000       ≤25               Low              Q1
9  Fatima   31   70000     26–35               Low              Q2 

Rows per equal-width bin: {'Low': 4, 'Mid': 3, 'High': 3}
Rows per quartile       : {'Q1': 3, 'Q2': 2, 'Q3': 2, 'Q4': 3}


### Common mistakes
Values outside `bins` become NaN (e.g. age 0 with `bins=[0, …]` because intervals are right-closed — use `include_lowest=True`) · `qcut` failing on many repeated values (`duplicates='drop'`) · forgetting that bins lose information.

---
<a id="sec-20"></a>
# 20. `explode()`

### What
Turns a column of **lists** (or list-like strings after splitting) into **one row per item**, repeating the other columns.

```python
df['Tags'] = df['Tags'].str.split(', ')
df.explode('Tags')
```

### When
Orders with several products in one cell, users with several skills/tags, movies with several genres → you need one row per item to count or group them.

In [29]:
skills = pd.DataFrame({'Name': ['Ali', 'Sara', 'Omar'], 'Skills': ['Python, SQL, Excel', 'Python, Tableau', 'Excel']})
skills['Skills'] = skills['Skills'].str.split(', ')           # text → list
exploded = skills.explode('Skills', ignore_index=True)       # one row per skill
print(exploded, '\n')
print('Most common skills:'); print(exploded['Skills'].value_counts())

   Name   Skills
0   Ali   Python
1   Ali      SQL
2   Ali    Excel
3  Sara   Python
4  Sara  Tableau
5  Omar    Excel 

Most common skills:
Skills
Python     2
Excel      2
SQL        1
Tableau    1
Name: count, dtype: int64


### Common mistakes
Exploding a **string** (nothing happens — split it into a list first) · forgetting the index repeats unless `ignore_index=True` · exploding two list columns separately (use `df.explode(['A', 'B'])` when their lists line up).

---
<a id="sec-21"></a>
# 21. Advanced Pandas Manipulation

<a id="sec-22"></a>
## 21.1 Reshaping: MultiIndex, `stack`, `unstack`, `melt`, `pivot`, `pivot_table`, `wide_to_long`

| Function | From → to | Simple meaning |
|---|---|---|
| **MultiIndex** | — | An index with several levels (e.g. City → Department) |
| `stack()` | wide → long | Move **columns** into the (inner) **index** |
| `unstack()` | long → wide | Move an (inner) **index** level into **columns** |
| `melt()` | wide → long | Turn columns into rows of `variable`/`value` (tidy data for plots/analysis) |
| `pivot()` | long → wide | Reshape **without aggregating** (fails if duplicates) |
| `pivot_table()` | long → wide | Reshape **with aggregation** (handles duplicates) |
| `wide_to_long()` | wide → long | For column names with a stub + number, e.g. `sales_2023`, `sales_2024` |

In [30]:
mi = emp.groupby(['City', 'Department'])['Salary'].mean()          # a Series with a 2-level MultiIndex
print('MultiIndex levels:', mi.index.names); print(mi.head(4), '\n')
print('Select one city from the MultiIndex:'); print(mi.loc['Lahore'], '\n')

wide = mi.unstack()                                                 # Department level → columns
print('unstack():'); print(wide, '\n')
print('stack() back to long:'); print(wide.stack().head(4))

MultiIndex levels: ['City', 'Department']
City       Department
Islamabad  Finance       110000.0
           Sales          70000.0
Karachi    HR             60000.0
           IT            104000.0
Name: Salary, dtype: float64 

Select one city from the MultiIndex:
Department
HR        55000.0
IT       102500.0
Sales     45000.0
Name: Salary, dtype: float64 

unstack():
Department   Finance       HR        IT    Sales
City                                            
Islamabad   110000.0      NaN       NaN  70000.0
Karachi          NaN  60000.0  104000.0      NaN
Lahore           NaN  55000.0  102500.0  45000.0 

stack() back to long:
City       Department
Islamabad  Finance       110000.0
           HR                 NaN
           IT                 NaN
           Sales          70000.0
dtype: float64


In [31]:
sales_wide = pd.DataFrame({'Store': ['A', 'B', 'C'], 'Jan': [100, 80, 120], 'Feb': [110, 95, 115], 'Mar': [130, 90, 140]})
long = sales_wide.melt(id_vars='Store', var_name='Month', value_name='Sales')    # wide → long
print('melt():'); print(long, '\n')
print('pivot() back to wide:'); print(long.pivot(index='Store', columns='Month', values='Sales'), '\n')

dup_long = pd.concat([long, pd.DataFrame({'Store': ['A'], 'Month': ['Jan'], 'Sales': [50]})])
try:
    dup_long.pivot(index='Store', columns='Month', values='Sales')
except ValueError as e:
    print('pivot() with duplicates →', str(e)[:60], '...')
print(dup_long.pivot_table(index='Store', columns='Month', values='Sales', aggfunc='sum'), '\n')

yearly = pd.DataFrame({'Store': ['A', 'B'], 'sales_2023': [500, 400], 'sales_2024': [650, 380], 'staff_2023': [5, 4], 'staff_2024': [6, 4]})
print('wide_to_long():')
print(pd.wide_to_long(yearly, stubnames=['sales', 'staff'], i='Store', j='Year', sep='_').reset_index())

melt():
  Store Month  Sales
0     A   Jan    100
1     B   Jan     80
2     C   Jan    120
3     A   Feb    110
4     B   Feb     95
5     C   Feb    115
6     A   Mar    130
7     B   Mar     90
8     C   Mar    140 

pivot() back to wide:
Month  Feb  Jan  Mar
Store               
A      110  100  130
B       95   80   90
C      115  120  140 



pivot() with duplicates → Index contains duplicate entries, cannot reshape ...
Month  Feb  Jan  Mar
Store               
A      110  150  130
B       95   80   90
C      115  120  140 

wide_to_long():


  Store  Year  sales  staff
0     A  2023    500      5
1     B  2023    400      4
2     A  2024    650      6
3     B  2024    380      4

<a id="sec-23"></a>
## 21.2 Conditional values: `where`, `mask`, `clip`

| Method | Keeps | Replaces |
|---|---|---|
| `s.where(cond, other)` | values where cond is **True** | the rest with `other` (NaN by default) |
| `s.mask(cond, other)` | values where cond is **False** | where cond is True |
| `s.clip(lower, upper)` | values inside the range | values outside → the nearest limit (capping outliers, Topic 5) |

In [32]:
sal = emp['Salary']
pd.DataFrame({'Salary': sal,
              'where >= 80k (else NaN)': sal.where(sal >= 80000),
              'mask > 100k → 100k': sal.mask(sal > 100000, 100000),
              'clip 50k–110k': sal.clip(50000, 110000)}).head(8)

,Salary,where >= 80k (else NaN),mask > 100k → 100k,clip 50k–110k
0,85000,85000.0,85000,85000
1,60000,NaN,60000,60000
2,120000,120000.0,100000,110000
3,90000,90000.0,90000,90000
4,98000,98000.0,98000,98000
5,55000,NaN,55000,55000
6,130000,130000.0,100000,110000
7,110000,110000.0,100000,110000


<a id="sec-24"></a>
## 21.3 Ranking and ordered calculations: `rank`, `shift`, `diff`, `pct_change`, cumulative functions

| Method | Meaning |
|---|---|
| `rank(ascending=False, method='dense')` | Position in sorted order (ties handled by `method`: `'average'`, `'min'`, `'dense'`, `'first'`) |
| `shift(1)` | Value from the **previous** row (lag) |
| `diff()` | Current − previous |
| `pct_change()` | (Current − previous) / previous |
| `cumsum()` / `cumprod()` | Running total / running product |
| `cummax()` / `cummin()` | Highest / lowest value so far |

⚠️ These depend on **row order** — sort first (usually by date), and use `groupby(...)` when rows belong to different entities.

In [33]:
monthly = pd.DataFrame({'Month': pd.period_range('2025-01', periods=8, freq='M'),
                        'Revenue': [100, 120, 115, 140, 160, 150, 175, 190],
                        'Growth_Factor': [1.00, 1.05, 0.98, 1.10, 1.02, 0.95, 1.08, 1.04]})
m = monthly.copy()
m['Rank'] = m['Revenue'].rank(ascending=False, method='dense').astype(int)
m['Prev_Month'] = m['Revenue'].shift(1)
m['Change'] = m['Revenue'].diff()
m['Pct_Change_%'] = (m['Revenue'].pct_change() * 100).round(1)
m['Cumulative'] = m['Revenue'].cumsum()
m['Best_So_Far'] = m['Revenue'].cummax()
m['Worst_So_Far'] = m['Revenue'].cummin()
m['Compounded_Index'] = m['Growth_Factor'].cumprod().round(3)
m

,Month,Revenue,Growth_Factor,Rank,Prev_Month,Change,Pct_Change_%,Cumulative,Best_So_Far,Worst_So_Far,Compounded_Index
0,2025-01,100,1.00,8,NaN,NaN,NaN,100,100,100,1.000
1,2025-02,120,1.05,6,100.0,20.0,20.0,220,120,100,1.050
2,2025-03,115,0.98,7,120.0,-5.0,-4.2,335,120,100,1.029
3,2025-04,140,1.10,5,115.0,25.0,21.7,475,140,100,1.132
4,2025-05,160,1.02,3,140.0,20.0,14.3,635,160,100,1.155
5,2025-06,150,0.95,4,160.0,-10.0,-6.2,785,160,100,1.097
6,2025-07,175,1.08,2,150.0,25.0,16.7,960,175,100,1.185
7,2025-08,190,1.04,1,175.0,15.0,8.6,1150,190,100,1.232


<a id="sec-25"></a>
## 21.4 Window functions: `rolling` and `expanding`

| | `rolling(3)` | `expanding()` |
|---|---|---|
| Window | The **last 3 rows** (moves along) | **All rows so far** (grows) |
| Example | 3-month moving average (smooths noise) | Average of all months to date |

In [34]:
w = monthly[['Month', 'Revenue']].copy()
w['Rolling_3_Mean'] = w['Revenue'].rolling(3).mean().round(1)
w['Rolling_3_Sum'] = w['Revenue'].rolling(3).sum()
w['Rolling_3_Mean_min1'] = w['Revenue'].rolling(3, min_periods=1).mean().round(1)   # allow partial windows at the start
w['Expanding_Mean'] = w['Revenue'].expanding().mean().round(1)
w['Expanding_Max'] = w['Revenue'].expanding().max()
w

,Month,Revenue,Rolling_3_Mean,Rolling_3_Sum,Rolling_3_Mean_min1,Expanding_Mean,Expanding_Max
0,2025-01,100,NaN,NaN,100.0,100.0,100.0
1,2025-02,120,NaN,NaN,110.0,110.0,120.0
2,2025-03,115,111.7,335.0,111.7,111.7,120.0
3,2025-04,140,125.0,375.0,125.0,118.8,140.0
4,2025-05,160,138.3,415.0,138.3,127.0,160.0
5,2025-06,150,150.0,450.0,150.0,130.8,160.0
6,2025-07,175,161.7,485.0,161.7,137.1,175.0
7,2025-08,190,171.7,515.0,171.7,143.8,190.0


### Common mistakes (Section 21)
- `pivot` on data with duplicate index/column pairs (use `pivot_table`).
- `melt` without `id_vars` (all columns get melted).
- `shift`/`diff`/`rolling` on unsorted data or across groups.
- Forgetting that `rolling(3)` gives NaN for the first 2 rows (`min_periods`).

---
<a id="sec-26"></a>
# 22. Performance

| Topic | Advice |
|---|---|
| **Vectorization** | Operate on whole columns (`df['a'] * df['b']`, `np.where`, `.str`, `.dt`) — runs in fast compiled code |
| **Loops** | `for` loops / `iterrows()` run Python code per row → slow; avoid for data transformations |
| **`apply()`** | Convenient but still a Python loop; prefer vectorized alternatives, especially `apply(axis=1)` |
| **Memory usage** | Check with `df.memory_usage(deep=True)` / `df.info(memory_usage='deep')` |
| **Efficient dtypes** | Downcast numbers (`int64` → `int32`/`int16`, `float64` → `float32`) when the range allows |
| **`category` dtype** | For text columns with **few distinct values** (city, gender, status): stores each label once + small integer codes → much less memory, faster groupby |
| **Avoid unnecessary copies** | Don't `.copy()` at every step; select only needed columns early; with Copy-on-Write (pandas 3), copies are made lazily only when you modify data |
| **Read less** | `pd.read_csv(..., usecols=[...], dtype={...})` |

In [35]:
n = 500_000
rng = np.random.default_rng(22)
big = pd.DataFrame({'city': rng.choice(['Lahore', 'Karachi', 'Islamabad', 'Multan'], n),
                    'qty': rng.integers(1, 50, n),
                    'price': rng.uniform(10, 5000, n)})

# 1) Loop vs vectorized
t0 = time.perf_counter()
loop_total = [q * p for q, p in zip(big['qty'].to_list()[:100_000], big['price'].to_list()[:100_000])]
t_loop = (time.perf_counter() - t0) * 5                                         # scaled to 500k rows
t0 = time.perf_counter(); vec_total = big['qty'] * big['price']; t_vec = time.perf_counter() - t0
print(f'Python loop (estimated for 500k rows): {t_loop:.3f} s | vectorized: {t_vec:.4f} s')

# 2) Memory: object/str vs category, int64 vs int8, float64 vs float32
before = big.memory_usage(deep=True).sum() / 1e6
opt = big.assign(city=big['city'].astype('category'), qty=big['qty'].astype('int8'), price=big['price'].astype('float32'))
after = opt.memory_usage(deep=True).sum() / 1e6
print(f'\nMemory before: {before:.1f} MB | after dtype optimisation: {after:.1f} MB')
print(opt.dtypes)

# 3) groupby speed on category vs text
t0 = time.perf_counter(); big.groupby('city')['price'].mean(); t_obj = time.perf_counter() - t0
t0 = time.perf_counter(); opt.groupby('city', observed=True)['price'].mean(); t_cat = time.perf_counter() - t0
print(f'\ngroupby on text: {t_obj:.4f} s | on category: {t_cat:.4f} s')

Python loop (estimated for 500k rows): 0.562 s | vectorized: 0.0009 s


Memory before: 15.5 MB | after dtype optimisation: 3.0 MB
city     category
qty          int8
price     float32
dtype: object

groupby on text: 0.0248 s | on category: 0.0063 s


### When not
- `category` for columns with **many unique values** (IDs, free text) — no savings and slower.
- Downcasting beyond the value range (`int8` holds only −128…127 → overflow!).
- Premature optimisation on tiny data — readability first.

### Common mistakes
`iterrows()` for calculations · converting to `category` and then comparing with strings that aren't categories (works, but watch for unused categories in groupby — use `observed=True`) · chained operations that repeatedly copy big DataFrames.

> ✍️ **Practice time:** Practice Problems (20), ✍️ Practice Problem 1, ✍️ Practice Problem 2, ✍️ Practice Problem 3, ✍️ Practice Problem 4, ✍️ Practice Problem 5 … (+16 more) — open [Topic_09_Practice_Solutions.ipynb](Topic_09_Practice_Solutions.ipynb) to solve it and see the solution.